Mldb 2111 allow select of more complex group by keys

steadtler opened the pull request.

guyd commented on server/bound_queries.cc:
```diff
@@ -1620,6 +1666,34 @@ BoundGroupByQuery(const SelectExpression & select,
     subSelect.reset(new BoundSelectQuery(subSelectExpr, from, alias, when, where, subOrderBy, calc, numBuckets));
 
+    std::vector<std::shared_ptr<ExpressionValueInfo> > groupInfo;
+    for (size_t c = 0; c < groupBy.clauses.size(); ++c) {
```
why not
```
for (const auto & boundCalc : subSelect->boundCalc)
```

steadtler replied:
Because the size of the calc vector is greater or equal to the size of the group by keys, since we also put aggregator's arguments in the calc vector, after the group by clauses.

guyd commented on sql/sql_expression_operations.cc:
```diff
@@ -4197,6 +4197,52 @@ wildcards() const
+GroupByKeyExpression::
+print() const
+{
+    return "GroupBy Key";
```
Don't we need to have the index at least?

guyd commented on sql/sql_expression_operations.cc:
```diff
@@ -4197,6 +4197,52 @@ wildcards() const
+GroupByKeyExpression::
+transform(const TransformArgs & transformArgs) const
+{
+    return make_shared<GroupByKeyExpression>(index);
```
Shouldn't we throw here that it is not implemented?

steadtler replied:
it is implemented, but we dont need to transform the args because it has no children expressions.

guyd commented on sql/sql_expression_operations.h:
```diff
@@ -599,4 +599,36 @@ struct SelectColumnExpression: public SqlRowExpression {
+
+    virtual std::string getType() const override
+    {
+        return "groupbykey";
```
`SelectColumnExpression` uses camel case with first lowercase.

guyd commented on testing/MLDB-2111-group-by-expression.py:
```diff
@@ -0,0 +1,127 @@
+
+    def test_groupby_expression_named(self):
+        res = mldb.query("""
+            SELECT x+1 as z FROM (SELECT x:1) GROUP BY x+1
```
Group by with a single row might not reveal all the issues.

steadtler replied:
Its good to have a variety. Some use the full dataset, some uses the select

guyd commented on testing/MLDB-2111-group-by-expression.py:
```diff
@@ -0,0 +1,127 @@
+    ## Rowname & rowHash in the select is different than in the group by
+    def test_groupby_rowname(self):
+        res = mldb.query("""
+            SELECT rowName() FROM ds GROUP BY rowName() LIMIT 1
```
Why the limit?  I would like to see all the grouped by rows.

steadtler commented on server/bound_queries.h:
```diff
@@ -173,13 +174,13 @@ struct BoundGroupByQuery {
 
     // Having Expression to resolve
-    const SqlExpression& having;
+    std::shared_ptr<SqlExpression> having;
```
Because we replace it and SqlExpression should only be allocated as shared_ptr

steadtler commented on sql/sql_expression_operations.cc:
```diff
@@ -4197,6 +4197,52 @@ wildcards() const
+                            + "' didn't return info");
+    }
+
+    auto outputInfo = getGroupbyKey.info->getConst(false);
```
We need the (group by) row to evaluate its constValue(). Its a bit tricky, so I put a comment to show its not an oversight.

steadtler commented:
@guyd 

I addressed most of the comments. Remember that this is not an optional paths, so all the other group by unit tests will still test regressions.

guyd commented on testing/MLDB-2111-group-by-expression.py:
```diff
@@ -123,5 +133,28 @@ def test_groupby_having(self):
+        msg = "variable 'x' must appear in the GROUP BY clause or be used in an aggregate function"
+        with self.assertRaisesRegexp(mldb_wrapper.ResponseException, msg):
+            res = mldb.query("""
+                SELECT x+1*3 FROM (SELECT x:1) GROUP BY 1+x
```
Here I would have liked to see `GROUP BY x+1`.

guyd commented on testing/MLDB-2111-group-by-expression.py:
```diff
@@ -123,5 +133,28 @@ def test_groupby_having(self):
 
+    def test_groupby_temporal(self):
+        res = mldb.query("""
+            select temporal_latest({*}) as * from ds2
```
I was thinking of `GROUP BY temporal_latest(x)`.
